lib/CL/devices/cpu_dbk/dbk_khr_jpeg_cpu.c

```diff
@@ -0,0 +1,217 @@
+{
+  cl_device_id dev = program->devices[dev_i];
+  cl_dbk_attributes_exp_jpeg_encode *attributes
+    = (cl_dbk_attributes_exp_jpeg_encode *)meta->builtin_kernel_attrs;
+  unsigned mem_id = dev->global_mem_id;
+  void *image = pocl_cpu_get_ptr (&arguments[0], mem_id);
+  void *jpeg = pocl_cpu_get_ptr (&arguments[1], mem_id);
+  uint64_t *jpeg_size = pocl_cpu_get_ptr (&arguments[2], mem_id);
+
+  jpeg_encode_state_t *state = kernel->data[dev_i];
+
+  int status;
+  status = tj3Compress8 (state->tj_handle, image, attributes->width, 0,
+                         attributes->height, TJPF_RGB, &(state->tj_buffer),
+                         &(state->jt_buffer_size));
+  if (status != 0)
+    {
+      POCL_MSG_ERR ("tj3Compress8 failed: %s (fatal? %d)\n",
+                    tj3GetErrorStr (state->tj_handle),
+                    tj3GetErrorCode (state->tj_handle));
+      return CL_OUT_OF_RESOURCES;
+    }
+  /* TODO remove this once the handle is configured with jpeg no realloc */
+  cl_mem jpeg_mem = *(cl_mem *)(arguments[0].value);
+  if (jpeg_mem->size < state->jt_buffer_size)
+    {
+      POCL_MSG_ERR ("pocl_cpu_execut_dbk_khr_jpeg_encode, "
+                    "compressed jpeg image is larger than destination.\n ");
+      assert (0 && "improper parameters");
+      return CL_OUT_OF_RESOURCES;
+    }
+  memcpy (jpeg, state->tj_buffer, state->jt_buffer_size);
+  *jpeg_size = state->jt_buffer_size;
+  return status;
+}
+
+void *
+pocl_cpu_init_dbk_khr_jpeg_encode (void const *attributes, int *status)
+{
+
+  jpeg_encode_state_t *state
+    = (jpeg_encode_state_t *)calloc (1, sizeof (jpeg_encode_state_t));
+  state->tj_handle = tj3Init (TJINIT_COMPRESS);
+  if (state->tj_handle == NULL)
+    {
+      const char *err_str = tj3GetErrorStr (state->tj_handle);
+      POCL_MSG_ERR ("create compress tj_handle failed: %s (fatal? %d)\n",
+                    err_str, tj3GetErrorCode (state->tj_handle));
+      *status = CL_OUT_OF_RESOURCES;
+      return NULL;
+    }
+
+  int jpeg_status;
+  cl_dbk_attributes_exp_jpeg_encode *jpeg_attr
+    = (cl_dbk_attributes_exp_jpeg_encode *)attributes;
+  jpeg_status = tj3Set (state->tj_handle, TJPARAM_QUALITY, jpeg_attr->quality);
+  if (jpeg_status != 0)
+    {
+      POCL_MSG_ERR ("could not set jpeg quality: %s (fatal? %d)\n",
+                    tj3GetErrorStr (state->tj_handle),
+                    tj3GetErrorCode (state->tj_handle));
+      *status = CL_OUT_OF_RESOURCES;
+      return NULL;
+    }
+
+  /* TODO: should this be configurable from the attributes? */
+  jpeg_status = tj3Set (state->tj_handle, TJPARAM_SUBSAMP, TJSAMP_420);
+  if (jpeg_status != 0)
+    {
+      POCL_MSG_ERR2 ("could not set jpeg subsampling", "%s",
+                     tj3GetErrorStr (state->tj_handle));
+      *status = CL_OUT_OF_RESOURCES;
+      return NULL;
+    }
+  /* todo set jpeg no_realloc */
+
+  return state;
+}
+
+int
+pocl_cpu_destroy_dbk_khr_jpeg_encode (void **kernel_data)
```

Comment:
Yeah. The whole module should be named after turbojpeg since this is turbojpeg impl. of the JPEG DBK. Let's not involve the device driver names here as we can integrate these also to custom devices etc.

Reply:
I was following the trend that these dbks are only implemented for cpu devices like basic and pthread, Just like common_utils.c is intended only for cpu devices. I'd imagine it'd give some confusion if non-cpu devices where using turbojpeg in the backend